File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/aas/impl/metamodel/hashmap/aas/submodelelement/SubmodelElement.java
package org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.submodelelement;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import org.eclipse.basyx.aas.api.metamodel.aas.qualifier.qualifiable.IConstraint;
import org.eclipse.basyx.aas.api.metamodel.aas.reference.IReference;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.ISubmodelElement;
import org.eclipse.basyx.aas.impl.metamodel.facades.HasDataSpecificationFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.HasKindFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.HasSemanticsFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.QualifiableFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.ReferableFacade;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.HasDataSpecification;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.HasSemantics;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.Referable;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.haskind.HasKind;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.qualifiable.Qualifiable;

public class SubmodelElement extends HashMap<String, Object> implements ISubmodelElement {
	private static final long serialVersionUID = 1L;

	public SubmodelElement() {
		putAll(new HasDataSpecification());
		putAll(new Referable());
		putAll(new Qualifiable());
		putAll(new HasSemantics());
		putAll(new HasKind());
	}

	public static SubmodelElement createAsFacade(Map<String, Object> obj) {
		SubmodelElement elem = new SubmodelElement();
		elem.putAll(obj);
		return elem;
	}

	@Override
	public Set<IReference> getDataSpecificationReferences() {
		return new HasDataSpecificationFacade(this).getDataSpecificationReferences();
	}

	public void setDataSpecificationReferences(Set<IReference> ref) {
		new HasDataSpecificationFacade(this).setDataSpecificationReferences(ref);
	}

	@Override
	public String getIdshort() {
		return new ReferableFacade(this).getIdshort();
	}

	@Override
	public String getCategory() {
		return new ReferableFacade(this).getCategory();
	}

	@Override
	public String getDescription() {
		return new ReferableFacade(this).getDescription();
	}

	@Override
	public IReference getParent() {
		return new ReferableFacade(this).getParent();
	}

	public void setIdshort(String idShort) {
		new ReferableFacade(this).setIdshort(idShort);
	}

	public void setCategory(String category) {
		new ReferableFacade(this).setCategory(category);
	}

	public void setDescription(String description) {
		new ReferableFacade(this).setDescription(description);
	}

	public void setParent(IReference obj) {
		new ReferableFacade(this).setParent(obj);
	}

	public void setQualifier(Set<IConstraint> qualifiers) {
		new QualifiableFacade(this).setQualifier(qualifiers);
	}

	@Override
	public Set<IConstraint> getQualifier() {
		return new QualifiableFacade(this).getQualifier();
	}

	@Override
	public IReference getSemanticId() {
		return new HasSemanticsFacade(this).getSemanticId();
	}

	public void setSemanticID(IReference ref) {
		new HasSemanticsFacade(this).setSemanticID(ref);
	}

	@Override
	public String getHasKindReference() {
		return new HasKindFacade(this).getHasKindReference();
	}

	public void setHasKindReference(String kind) {
		new HasKindFacade(this).setHasKindReference(kind);
	}

	@Override
	public String getId() {
		return (String) get(Referable.IDSHORT);
	}

	@Override
	public void setId(String id) {
		put(Referable.IDSHORT, id);
	}

}

File: sdks/c++/basys.sdk.cc/src/aas/impl/metamodel/hashmap/aas/submodelelement/SubmodelElement.cpp
/*
 * SubmodelElement.cpp
 *
 *      Author: wendel
 */

#include "SubmodelElement.h"

namespace basyx {
namespace submodel {
namespace metamodel {
namespace map {
namespace submodelelement {

using namespace aas;

SubmodelElement::SubmodelElement() :
  description {"", ""}
{}

//SubmodelElement::SubmodelElement(const basyx::objectMap_t & submodelElementMap)
//{
//
//}

basyx::specificCollection_t<reference::IReference> SubmodelElement::getDataSpecificationReferences() const
{
  return this->dataSpecificationReferences;
}

std::string SubmodelElement::getIdShort() const
{
  return this->idShort;
}

std::string SubmodelElement::getCategory() const
{
  return this->category;
}

aas::qualifier::impl::Description SubmodelElement::getDescription() const
{
  return this->description;
}

std::shared_ptr<reference::IReference> SubmodelElement::getParent() const
{
  return this->parent;
}

basyx::specificCollection_t<aas::qualifier::qualifiable::IConstraint> SubmodelElement::getQualifier() const
{
  return this->qualifier;
}

std::shared_ptr<reference::IReference> SubmodelElement::getSemanticId() const
{
  return this->semanticId;
}

qualifier::haskind::Kind SubmodelElement::getHasKindReference() const
{
  return this->hasKindReference;
}

void SubmodelElement::setDataSpecificationReferences(const basyx::specificCollection_t<aas::reference::IReference> & dataSpecificationReferences)
{
  this->dataSpecificationReferences = dataSpecificationReferences;
}

void SubmodelElement::setIdShort(const std::string & idShort)
{
  this->idShort = idShort;
}

void SubmodelElement::setCategory(const std::string & category)
{
  this->category = category;
}

void SubmodelElement::setDescription(const aas::qualifier::impl::Description & description)
{
  this->description = description;
}

void SubmodelElement::setParent(const std::shared_ptr<aas::reference::IReference> & parent)
{
  this->parent = parent;
}

void SubmodelElement::setQualifier(const basyx::specificCollection_t<aas::qualifier::qualifiable::IConstraint> & qualifier)
{
  this->qualifier = qualifier;
}

void SubmodelElement::setSemanticId(const std::shared_ptr<aas::reference::IReference> & semanticId)
{
  this->semanticId = semanticId;
}

void SubmodelElement::setHasKindReference(const qualifier::haskind::Kind & hasKindReference)
{
  this->hasKindReference = hasKindReference;
}

}
}
}
}
}
